### PR TITLE
[SPARK-32210][CORE] Fix NegativeArraySizeException in MapOutputTracker with large spark.default.parallelism

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
### What changes were proposed in this pull request?
The current `MapOutputTracker` class may throw `NegativeArraySizeException` with a large number of partitions. Within the serializeOutputStatuses() method, it is trying to compress an array of mapStatuses and outputting the binary data into (Apache)ByteArrayOutputStream . Inside the (Apache)ByteArrayOutputStream.toByteArray(), negative index exception happens because the index is int and overflows (2GB limit) when the output binary size is too large.

This PR proposes two high-level ideas:
  1. Use `org.apache.spark.util.io.ChunkedByteBufferOutputStream`, which has a way to output the underlying buffer as `Array[Array[Byte]]`.
  2. Change the signatures from `Array[Byte]` to `Array[Array[Byte]]` in order to handle over 2GB compressed data.

### Why are the changes needed?
This issue seems to be missed out in the earlier effort of addressing 2GB limitations [SPARK-6235](https://issues.apache.org/jira/browse/SPARK-6235)

Without this fix, `spark.default.parallelism` needs to be kept at the low number. The drawback of setting smaller spark.default.parallelism is that it requires more executor memory (more data per partition). Setting `spark.io.compression.zstd.level` to higher number (default 1) hardly helps.

That essentially means we have the data size limit that for shuffling and does not scale.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Passed existing tests
```
build/sbt "core/testOnly org.apache.spark.MapOutputTrackerSuite"
```
Also added a new unit test
```
build/sbt "core/testOnly org.apache.spark.MapOutputTrackerSuite  -- -z SPARK-32210"
```
Ran the benchmark using GitHub Actions and didn't not observe any performance penalties. The results are attached in this PR
```
core/benchmarks/MapStatusesSerDeserBenchmark-jdk11-results.txt
core/benchmarks/MapStatusesSerDeserBenchmark-results.txt
```

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1416,6 +1416,17 @@ package object config {
       .stringConf
       .createWithDefault("zstd")
 
+  private[spark] val MAP_STATUS_OUTPUT_CHUNK_SIZE =
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1419]: Do we really need to make this a config option?

**Review by srowen — COMMENTED** [review]
Seems reasonable

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @kazuyukitanimura .

BTW, could you take a look at the UT failures?
- https://github.com/kazuyukitanimura/spark/runs/3309290960?check_suite_focus=true

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @Ngone51 too FYI

**Comment by dbtsai** [comment]
Jenkins, add to whitelist

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1419]: According to @srowen and @mridulm 's advice, shall we remove this, @kazuyukitanimura ?

**Comment by dongjoon-hyun** [comment]
Retest this please

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
Comment (kazuyukitanimura) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Thanks, removed.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 2]
```diff
@@ -1335,22 +1334,23 @@ private[spark] object MapOutputTracker extends Logging {
...
+    // create a nested Array so that it can handle over 2GB serialized data
+    val arr = out.toChunkedByteBuffer.getChunks().map(_.array())
+    val arrSize = out.size
+    if (arrSize >= minBroadcastSize) {
       // Use broadcast instead.
       // Important arr(0) is the tag == DIRECT, ignore that while deserializing !
       val bcast = broadcastManager.newBroadcast(arr, isLocal)
-      // toByteArray creates copy, so we can reuse out
-      out.reset()
+      val out = new ChunkedByteBufferOutputStream(chunkSize, ByteBuffer.allocate _)
```
[the remark this answers is not in the document]
Comment (kazuyukitanimura) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: Good point! updated.

**Review by mridulm — APPROVED** [review]
LGTM, thanks for working on this @kazuyukitanimura !

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @kazuyukitanimura and all.
Merged to master/3.2.

**Comment by dongjoon-hyun** [comment]
cc @gengliangwang

File: core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala [hunk 3]
```diff
@@ -664,4 +664,74 @@ class MapOutputTrackerSuite extends SparkFunSuite with LocalSparkContext {
...
+      val result = MapOutputTracker.deserializeOutputStatuses(serializedMapStatus, sc.getConf)
+      assert(result.length == numMaps)
+
+      tracker.unregisterShuffle(shuffleId)
+      tracker.stop()
+    }
+  }
+
+  ignore("SPARK-32210: serialize and deserialize over 2GB compressed mapStatuses") {
+    // This test requires 8GB heap memory settings
```
Comment (pan3793) [inline core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:704]: How can I run it in local? 
Still encounter `sbt.ForkMain$ForkError: java.lang.OutOfMemoryError: Java heap space` with `SBT_OPTS="-Xmx16G"`

Comment (kazuyukitanimura) [reply core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:704]: @pan3793 `project/SparkBuild.scala` is hard-coding the `Xmx` sizes unfortunately. You may need to manually change them, something like https://github.com/apache/spark/pull/34214/files
If you are using 8G should work. If you are using JDK 11 or 17, you may need more than 8G

**Pull request closed without merging**